### ISSUE TITLE
Metrics Undelete - Remove tombstones

### ISSUE THREAD

**Issue by roidelapluie** [prometheus/prometheus#7567]
## Proposal
**Use case. Why is this important?**

When you accidentally delete too much data and you notice directly, it would be great to be able to "undelete" - aka remove the tombstones entries you have just created.

How could it be done?

When you add a tombstone, the API would return a list of tombstones ID's - one per block? And then another API endpoint would be able to remove those tombstones.

**Comment by bwplotka** [comment]
cc @metalmatze  @Harshitha1234 who is working on the same thing on Thanos side, undelete were our aim as well.


> How could it be done?
> When you add a tombstone, the API would return a list of tombstones ID's - one per block? And then another API endpoint would be able to remove those tombstones.

With the current approach it's impossible. All tombstones are merged into one in non reversible manner. You could only potentially remove all tombstones for given block and only if you are lucky that this tombstones was not already processed.

This is why with @metalmatze and @Harshitha1234 we were exploring different formats of tombstones, like log based ones, that would allow us to remove individual tombstone appends. On top of that we planned to have time based tombstones - so each log would be applied (processed in irreversible manner) only after certain safety time (e.g `1d`). We were thinking to do that for Thanos / Cortex side (cc @pracucci ) but your issue indicate such work would be nice on Prometheus side as well! :hugs:

**Comment by metalmatze** [comment]
Our idea was to have an append-only log of tombstones where we add all the tombstones to.
If a user happens to delete too much they can un-delete a tombstone by creating a special tombstone that would mask the referencing tombstone as obsolete. Upon writing the new block the tombstones would be all read and the final state of tombstones would be seen in-memory with the deleted ones gone from that slides/map.

**Comment by brian-brazil** [comment]
> All tombstones are merged into one in non reversible manner.

This is quite important for performance, some users over-use deletions.

If this were added do you really need to be able to undo individual API calls, or in an emergency get rid of all tombstones and then re-add the ones you actually wanted?

**Comment by brian-brazil** [comment]
Also, users often want to immediately reduce the size of their tsdb after bad data has been ingested. Being forced to wait a day doesn't seem great there.

**Comment by bwplotka** [comment]
> Also, users often want to immediately reduce the size of their tsdb after bad data has been ingested. Being forced to wait a day doesn't seem great there.

Sure, that would be optional.

> This is quite important for performance, some users over-use deletions.
> If this were added do you really need to be able to undo individual API calls, or in an emergency get rid of all tombstones and then re-add the ones you actually wanted?

Ideally since we aim for multi tenant system and softtenancy in single block I believe individual API calls. cc @pracucci @brancz 

For Prometheus I agree, just full delete would be an option

**Comment by brian-brazil** [comment]
It sounds like the requirements for Thanos would be quite different to those for Prometheus, as Prometheus is single tenant.

**Comment by roidelapluie** [comment]
I would personnally be happy with simply either an undelete of everything that can still be saved. This is meant to be a last-resort thing.

**Comment by Harshitha1234** [comment]
> Also, users often want to immediately reduce the size of their tsdb after bad data has been ingested. Being forced to wait a day doesn't seem great there.

How about maybe having an option for the user to enter the timestamp until which the tombstones file needs to be preserved, this could help solve this case and as @bwplotka said we can have `1d` as certain default safety time. This extends a user the priviledge to have the tombstones and data safe until he actually wants the data to  be deleted.

**Comment by brian-brazil** [comment]
That sounds like a fair bit of complication for the sake of what is possibly going to be only friction for the user.

**Comment by Harshitha1234** [comment]
Okay.
> I would personnally be happy with simply either an undelete of everything that can still be saved. This is meant to be a last-resort thing.

So looks like undelete all would be the best option :).

**Comment by krajorama** [comment]
Hello from the bug scrub: there seems to be no further user request associated with this issue. Closing for no activity. Please follow-up if you want to work on the requirements and design of such feature.
